lib/collection/src/grouping/types.rs

```diff
@@ -0,0 +1,89 @@
+use std::hash::Hash;
+use std::ops::{Deref, DerefMut};
+
+use segment::types::ScoredPoint;
+use serde_json::Value;
+use AggregatorError::BadKeyType;
+
+#[derive(PartialEq, Debug)]
+pub(super) enum AggregatorError {
+    AllGroupsFull,
+    BadKeyType,
+    KeyNotFound,
+    GroupFull,
+    EnoughGroups,
+}
+
+/// Abstraction over serde_json::Value to be used as a key in a HashMap/HashSet
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub(super) struct GroupKey(pub serde_json::Value);
+
+impl TryFrom<serde_json::Value> for GroupKey {
+    type Error = AggregatorError;
+
+    /// Only allows Strings and Numbers to be converted into GroupKey
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        match value {
+            serde_json::Value::String(_) | serde_json::Value::Number(_) => Ok(Self(value)),
+            _ => Err(BadKeyType),
+        }
+    }
+}
+
+impl Hash for GroupKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match &self.0 {
+            Value::Number(n) => n.hash(state),
+            Value::String(s) => s.hash(state),
+            _ => unreachable!("GroupKey should only be a number or a string"),
+        }
+    }
+}
+
+/// Abstraction over ScoredPoint to be used in a HashSet
+#[derive(Eq, Debug, PartialEq, Clone)]
+pub(super) struct HashablePoint(pub ScoredPoint);
```

Comment:
I actually need to be able to hash points with vectors and payloads in the final stage of the `group_by()` to avoid doing extra loops in the hydration part. Btw, I refactored the `hydrate_from` into the `Group` implementation.

tldr: I did the inner ScoredPoint private